File: django_rest/src/main/java/com/kuky/restframework/RetrofitApiManager.java
package com.kuky.restframework;

import android.support.annotation.NonNull;
import android.util.Log;

import java.io.IOException;
import java.util.HashMap;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @author kuky
 * @description
 */
public class RetrofitApiManager {
    private static final String TAG = "RetrofitApiManager";
    private static final int CONN_TIME = 5;
    private static final int READ_TIME = 10;
    private static final int WRITE_TIME = 30;
    private static Retrofit mRetrofit;

    private RetrofitApiManager() {

    }

    public static Retrofit provideClient(String baseUrl, HashMap<String, String> headers) {
        if (mRetrofit == null) {
            synchronized (RetrofitApiManager.class) {
                if (mRetrofit == null) {
                    mRetrofit = new Retrofit.Builder()
                            .baseUrl(baseUrl)
                            .client(genericOkClient(headers))
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .addConverterFactory(GsonConverterFactory.create())
                            .build();
                }
            }
        }
        return mRetrofit;
    }

    /**
     * 创建 OkHttpClient 实例
     *
     * @param headers 添加头部信息
     * @return
     */
    private static OkHttpClient genericOkClient(final HashMap<String, String> headers) {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor(new HttpLoggingInterceptor.Logger() {
            @Override
            public void log(@NonNull String message) {
                Log.i(RetrofitApiManager.class.getSimpleName(), "Interceptor message =>" + message);
            }
        });

        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        return new OkHttpClient.Builder()
                .connectTimeout(CONN_TIME, TimeUnit.SECONDS)
                .readTimeout(READ_TIME, TimeUnit.SECONDS)
                .writeTimeout(WRITE_TIME, TimeUnit.SECONDS)
                .addNetworkInterceptor(loggingInterceptor)
                .addInterceptor(new Interceptor() {
                    @Override
                    public Response intercept(Chain chain) throws IOException {
                        Request.Builder builder = chain.request().newBuilder();
                        if (headers != null) {
                            for (String key : headers.keySet()) {
                                builder.addHeader(key, headers.get(key));
                            }
                        }
                        return chain.proceed(builder.build());
                    }
                }).build();
    }
}

File: base-kt-module/src/main/java/com/kuky/base_kt_module/RetrofitApiManager.kt
package com.kuky.base_kt_module

import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

/**
 * @author kuky.
 * @description
 */
object RetrofitApiManager {
    private const val TAG = "RetrofitApiManager"
    private const val CONN_TIME = 5
    private const val READ_TIME = 10
    private const val WRITE_TIME = 30
    private var mRetrofit: Retrofit? = null

    fun provideClient(baseUrl: String): Retrofit {
        if (mRetrofit == null) {
            mRetrofit = Retrofit.Builder()
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .addConverterFactory(GsonConverterFactory.create())
                    .client(genericOkClient())
                    .build()
        }

        return mRetrofit!!
    }

    private fun genericOkClient(): OkHttpClient {
        val httpLoggingInterceptor = HttpLoggingInterceptor {
            HttpLoggingInterceptor.Logger { message -> Log.i(TAG, message) }
        }

        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY

        return OkHttpClient.Builder()
                .connectTimeout(CONN_TIME.toLong(), TimeUnit.SECONDS)
                .readTimeout(READ_TIME.toLong(), TimeUnit.SECONDS)
                .writeTimeout(WRITE_TIME.toLong(), TimeUnit.SECONDS)
                .addInterceptor(httpLoggingInterceptor)
                .build()
    }
}